File: unix/cat.lua
local p = require "checkedposix"

-- NOTE: obviously you don't really want to do that,
-- allocations will kill you...

local STDIN_FILENO = p.fileno(io.stdin)
local STDOUT_FILENO = p.fileno(io.stdout)
local CHUNK_SZ = 512

local chunk, sz

while true do
    chunk = p.read(STDIN_FILENO, 512)
    if chunk == "" then break end;
    sz = p.write(STDOUT_FILENO, chunk); assert(sz == #chunk)
end

File: unix/cat.c
#include "usual.h"

#define CHUNK_SZ 512

int main(int argc, char **argv) {

    char buf[CHUNK_SZ];
    int r, w;

    for(;;) {
        r = read(STDIN_FILENO, buf, sizeof(buf));
        if (r == 0) break;
        assert (r > 0);
        w = write(STDOUT_FILENO, buf, r);
        assert (w == r);
    }

    return EXIT_SUCCESS;
}
